dcpy/models/lifecycle/ingest.py

```diff
@@ -38,55 +39,92 @@ class ScriptSource(BaseModel, extra="forbid"):
 )
 
 
-class PreprocessingStep(BaseModel):
+class PreprocessingStep(SortedSerializedBase):
     name: str
     args: dict[str, Any] = {}
     # mode allows for certain preprocessing steps only to be run if specified at runtime
     mode: str | None = None
 
 
-class Template(BaseModel, extra="forbid", arbitrary_types_allowed=True):
-    """Definition of a dataset for ingestion/processing/archiving in edm-recipes"""
+class DatasetAttributes(SortedSerializedBase):
+    name: str | None = None
+    description: str | None = None
+    url: str | None = None
+    custom: dict | None = None
 
-    id: str
+    _head_sort_order = ["name", "description", "url"]
+
+
+class ArchivalMetadata(SortedSerializedBase):
+    archival_timestamp: datetime
+    check_timestamps: list[datetime] = []
+    raw_filename: str
     acl: recipes.ValidAclValues
 
-    target_crs: str | None = None
 
-    ## these two fields might merge to "source" or something equivalent at some point
-    ## for now, they are distinct so that they can be worked on separately
-    ## when implemented, "None" should not be valid type
+class Ingestion(SortedSerializedBase):
+    target_crs: str | None = None
     source: Source
     file_format: file.Format
-
+    processing_mode: str | None = None
     processing_steps: list[PreprocessingStep] = []
 
+
+class Column(SortedSerializedBase):
+    id: str
+    data_type: Literal["text", "integer", "decimal", "geometry", "bool", "datetime"]
+    description: str | None = None
+
+
+class Template(BaseModel, extra="forbid"):
+    """Definition of a dataset for ingestion/processing/archiving in edm-recipes"""
+
+    id: str
+    acl: recipes.ValidAclValues
+
+    attributes: DatasetAttributes | None = None
+    ingestion: Ingestion
+    columns: list[Column] = []
+
     ## this is the original library template, included just for reference while we build out our new templates
     library_dataset: library.DatasetDefinition | None = None
 
+    @property
+    def has_geom(self):
+        match self.ingestion.file_format:
+            case file.Shapefile() | file.Geodatabase() | file.GeoJson():
+                return True
+            case file.Csv() | file.Xlsx() | file.Json() as format:
+                return format.geometry is not None
+
 
-class Config(BaseModel, extra="forbid", arbitrary_types_allowed=True):
+class Config(SortedSerializedBase, extra="forbid"):
     """
     Computed Template of ingest dataset
     Stored in config.json in edm-recipes/raw_datasets and edm-recipes/datasets
     """
 
     id: str = Field(validation_alias=AliasChoices("id", "name"))
     version: str
-    archival_timestamp: datetime
-    check_timestamps: list[datetime] = []
-    raw_filename: str
-    acl: recipes.ValidAclValues
-
-    target_crs: str | None = None
-
-    source: Source
-    file_format: file.Format
-    processing_mode: str | None = None
-    processing_steps: list[PreprocessingStep] = []
+    crs: str | None = None
```

Comment:
is this attribute here to accommodate the old template?

Reply:
Nothing in this PR is really to accommodate anything old. These config objects are meant for reading data about an archived dataset, so with that I thought that `crs` was worthy of being a top-level field since it's quite relevant to how we use the data. Given that it's also stored in the parquet metadata, this is maybe up for debate

Reply:
I see. My thought is that crs should "live" in one place under ingestion to avoid 1) duplication and 2) any confusion whether`crs` should or shouldn't be specified in both places.